Horizontal toolbox layout with positioning at start or end.

rachel-fenichel opened the pull request:
- Horizontal toolbox
- Positioning at start or end
- Playground with various combinations of layout, position, and rtl.
- Known bug with transhcan positioning with "end" orientation
- Known bug with scrolling in RTL horizontal flyout.

rachel-fenichel commented:
@picklesrus

picklesrus commented on core/flyout.js:
```diff
@@ -209,44 +232,77 @@ Blockly.Flyout.prototype.getMetrics_ = function() {
     viewHeight: viewHeight,
     viewWidth: viewWidth,
-    contentHeight: (optionBox.height + optionBox.y) * this.workspace_.scale,
+    contentHeight: (optionBox.height) * this.workspace_.scale,
```
why does the contentHeight calculation change? was optionBox.y always 0 before?

rachel-fenichel replied:
There's now a contentTop and contentLeft as well as contentHeight.  I think optionBox.y was generally 0 before, and that it didn't actually work if it was nonzero.

picklesrus commented on core/flyout.js:
```diff
@@ -209,44 +232,77 @@ Blockly.Flyout.prototype.getMetrics_ = function() {
+    }
   }
-  this.workspace_.translate(0, this.workspace_.scrollY + metrics.absoluteTop);
+  var translateX = this.horizontalLayout_ && this.RTL ?
```
I think I can follow most of these calculations in my head except this one.  Is it possible to describe this in a comment?

NeilFraser commented on core/css.js:
```diff
@@ -423,6 +423,16 @@ Blockly.Css.CONTENT = [
 
+  '.blocklyHorizontalTree {',
+    'float: left;',
+    'margin: 1px 5px 8px 0px;',
```
Google's CSS style guide says to just use 0 instead of 0px.

NeilFraser commented on core/css.js:
```diff
@@ -423,6 +423,16 @@ Blockly.Css.CONTENT = [
+
+  '.blocklyHorizontalTreeRtl {',
+    'float: right;',
+    'margin: 1px 0px 8px 5px;',
```
0

NeilFraser commented on core/css.js:
```diff
@@ -437,6 +447,14 @@ Blockly.Css.CONTENT = [
 
+  '.blocklyTreeSeparatorHorizontal {',
+    'border-right: solid #e5e5e5 1px;',
+    'width: 0px;',
```
0

NeilFraser replied:
Missed a 0.

NeilFraser commented on core/flyout.js:
```diff
@@ -209,44 +232,77 @@ Blockly.Flyout.prototype.getMetrics_ = function() {
   }
-  return {
+
+  var absoluteTop = this.verticalOffset_ + this.SCROLLBAR_PADDING
```
;

NeilFraser commented on core/flyout.js:
```diff
@@ -209,44 +232,77 @@ Blockly.Flyout.prototype.getMetrics_ = function() {
-        -metrics.contentHeight * yRatio.y - metrics.contentTop;
+        -metrics.contentHeight * xyRatio.y - metrics.contentTop;
+  } else if (this.horizontalLayout_ && goog.isNumber(xyRatio.x)) {
+    if (this.RTL) {
```
Suggest:
this.workspace_.scrollX = -metrics.contentWidth * xyRatio.x + (this.RTL ? metrics.contentLeft : -metrics.contentLeft);

NeilFraser commented on core/flyout.js:
```diff
@@ -259,47 +315,144 @@ Blockly.Flyout.prototype.position = function() {
+ * @private
+ */
+Blockly.Flyout.prototype.setBackgroundPathVertical_ = function(width, height) {
+    var atRight = this.toolboxPosition_ == Blockly.TOOLBOX_AT_RIGHT;
```
Too much indent.

NeilFraser commented on core/flyout.js:
```diff
@@ -259,47 +315,144 @@ Blockly.Flyout.prototype.position = function() {
+ * @private
+ */
+Blockly.Flyout.prototype.setBackgroundPathHorizontal_ = function(width, height) {
+    var atTop = this.toolboxPosition_ == Blockly.TOOLBOX_AT_TOP;
```
Too much indent.

NeilFraser commented on core/flyout.js:
```diff
@@ -259,47 +315,144 @@ Blockly.Flyout.prototype.position = function() {
+  var path = ['M 0,' + (atTop ? 0 : this.CORNER_RADIUS)];
+
+  if (atTop) {
+    // top
```
Top.  To make consistent with properly capitalized and punctuated comments in previous function.  Fix throughout this function.

NeilFraser commented on core/flyout.js:
```diff
@@ -259,47 +315,144 @@ Blockly.Flyout.prototype.position = function() {
  */
 Blockly.Flyout.prototype.scrollToStart = function() {
-  this.scrollbar_.set(0);
+  this.scrollbar_.set((this.horizontalLayout_ && this.RTL) ? 1000000000 : 0);
```
If Infinity works, that would be better.

NeilFraser commented on core/flyout.js:
```diff
@@ -473,51 +627,31 @@ Blockly.Flyout.prototype.show = function(xmlList) {
- * Compute width of flyout.  Position button under each block.
- * For RTL: Lay out the blocks right-aligned.
+ * Add listeners to a block that has been added to the flyout.
+ * @param {Element} root The root node of the SVG group the block is in.
```
{!Element}

NeilFraser commented on core/flyout.js:
```diff
@@ -579,13 +713,23 @@ Blockly.Flyout.prototype.onMouseDown_ = function(e) {
+    var metrics = this.getMetrics_();
+    var x = metrics.viewLeft - dx;
+    x = Math.min(x, metrics.contentWidth - metrics.viewWidth);
+    x = Math.max(x, 0);
```
Try using goog.math.clamp.  Here and once below.

rachel-fenichel replied:
[Give 'im the clamps!](http://imgur.com/gallery/fD5R7Xj)

NeilFraser commented on core/flyout.js:
```diff
@@ -708,13 +852,24 @@ Blockly.Flyout.prototype.getClientRect = function() {
+
+  if (this.toolboxPosition_ == Blockly.TOOLBOX_AT_TOP) {
+    return new goog.math.Rect(-BIG_NUM, y - BIG_NUM, BIG_NUM * 2,
+      BIG_NUM + height);
```
Indent more.  Here and thrice below.

NeilFraser commented on core/flyout.js:
```diff
@@ -742,3 +897,93 @@ Blockly.Flyout.terminateDrag_ = function() {
+  var flyoutHeight = 0;
+  var margin = this.CORNER_RADIUS;
+  var blocks = this.workspace_.getTopBlocks(false);
+  for (var x = 0, block; block = blocks[x]; x++) {
```
We prefer i to x, unless dealing with horizontal coordinates.  Blockly has a lot of x and y, so let's not use x to mean something else.

NeilFraser commented on core/flyout.js:
```diff
@@ -742,3 +897,93 @@ Blockly.Flyout.terminateDrag_ = function() {
+  var blocks = this.workspace_.getTopBlocks(false);
+  for (var x = 0, block; block = blocks[x]; x++) {
+    var height = block.getHeightWidth().height;
+    flyoutHeight = Math.max(flyoutHeight, height);
```
These two fit on one line.  No need for a temp variable.

NeilFraser commented on core/flyout.js:
```diff
@@ -742,3 +897,93 @@ Blockly.Flyout.terminateDrag_ = function() {
+  flyoutHeight *= this.workspace_.scale;
+  flyoutHeight += margin * 1.5 + Blockly.Scrollbar.scrollbarThickness;
+  if (this.height_ != flyoutHeight) {
+    for (var x = 0, block; block = blocks[x]; x++) {
```
i

NeilFraser commented on core/flyout.js:
```diff
@@ -742,3 +897,93 @@ Blockly.Flyout.terminateDrag_ = function() {
+  var flyoutWidth = 0;
+  var margin = this.CORNER_RADIUS;
+  var blocks = this.workspace_.getTopBlocks(false);
+  for (var x = 0, block; block = blocks[x]; x++) {
```
i

NeilFraser commented on core/flyout.js:
```diff
@@ -742,3 +897,93 @@ Blockly.Flyout.terminateDrag_ = function() {
+  flyoutWidth *= this.workspace_.scale;
+  flyoutWidth += margin * 1.5 + Blockly.Scrollbar.scrollbarThickness;
+  if (this.width_ != flyoutWidth) {
+    for (var x = 0, block; block = blocks[x]; x++) {
```
i

NeilFraser commented on core/toolbox.js:
```diff
@@ -313,16 +387,26 @@ Blockly.Toolbox.prototype.getClientRect = function() {
+  } else if (this.toolboxPosition == Blockly.TOOLBOX_AT_RIGHT) {
+    return new goog.math.Rect(x, -BIG_NUM, BIG_NUM + width, 2 * BIG_NUM);
+  } else if (this.toolboxPosition == Blockly.TOOLBOX_AT_TOP) {
+      return new goog.math.Rect(-BIG_NUM, -BIG_NUM, 2 * BIG_NUM,
```
Bad indent here and next three lines.

NeilFraser commented:
Looks good, if you'd like to address the comments and resolve the conflicts I'd like to torture test the code a bit.

rachel-fenichel commented:
It still has a significant problem with scrolling in the horizontal toolbox in RTL mode.  @picklesrus tried to fix it and is as confused as I am. 

Will address comments + fix conflicts shortly.

rachel-fenichel commented:
Comments addressed + conflicts resolved.  PTAL.

NeilFraser commented on core/flyout.js:
```diff
@@ -573,18 +703,26 @@ Blockly.Flyout.prototype.onMouseDown_ = function(e) {
+    this.startDragMouseY_ = e.clientY;
+    var metrics = this.getMetrics_();
+    var y = metrics.viewTop - dy;
+    y = goog.math.clamp(y, o, metrics.contentHeight - metrics.viewHeight);
```
It is terrifying that the typo on this line works.

NeilFraser commented:
Why are the categories all reversed in the regular playground?

Scrollbar position on flyout in {horizontalLayout: true, toolboxPosition: 'start'} is too high (in the case where there are too many blocks for the flyout to show).  The other three modes have the flyout's scrollbar be right at the edge.

NeilFraser commented:
Horizontal layout; toolbox at start, RTL:  The flyout's scrollbar scrolls in the wrong direction (left should be right).

NeilFraser commented:
Horizontal layout; toolbox at end; zoom in.  When a block is created, it is offset from the mouse cursor.

rachel-fenichel commented:
Categories reversed in vertical mode: whoops.  Fixed.

Scrollbar positioning: fixed.

Zoom: is that only for horizontal toolbox at end, and not for vertical toolbox at end?

Flyout scrolling in horizontal RTL: I don't know.

picklesrus commented on core/blockly.js:
```diff
@@ -406,14 +406,25 @@ Blockly.hideChaff = function(opt_allowToolbox) {
   var svgSize = Blockly.svgSize(this.getParentSvg());
   if (this.toolbox_) {
-    svgSize.width -= this.toolbox_.width;
+    // If the toolbox is at the bottom it's laid out separately from the main
```
I'm slightly confused by the comment and code.  You're checking for TOP, but the comment says bottom.  But I was expecting it to be TOP or BOTTOM?

rachel-fenichel replied:
Guess the comment's not clear enough, then.

In the case that the toolbox is at the top it seems to overlap the main workspace and therefore I need to subtract its height.  Also in the left and right cases.  But in the case where it's at the bottom it doesn't overlap, so I don't need to subtract its height.

...but if that's the case, then the else should be an else if, covering left and right, so that nothing is subtracted if it's at bottom.

picklesrus replied:
sorry to be a pain, but why doesn't it overlap on the bottom but it does on the top?  Just trying to understand the different code paths, why we have them, and what things need to be manually tested when this code changes.

picklesrus commented on core/css.js:
```diff
@@ -440,10 +450,18 @@ Blockly.Css.CONTENT = [
+  '.blocklyTreeSeparatorHorizontal {',
+    'border-right: solid #e5e5e5 1px;',
+    'width: 0;',
+    'padding: 5px 0;',
```
just curious - why the mix of padding and margin and not just either padding all around or margin all around?

picklesrus commented on core/flyout.js:
```diff
@@ -209,46 +250,74 @@ Blockly.Flyout.prototype.getMetrics_ = function() {
+    var optionBox = {height: 0, y: 0, width: 0, x: 0};
+  }
+
+  var absoluteTop = this.SCROLLBAR_PADDING;
```
I know it was before, so no objections, but do you know why is absoluteTop= scrollbar_padding in the vertical case? I was expecting 0 or some constant unrelated to scrolling. Something like Flyout.MARGIN or Flyout.CORNER_RADIUS

rachel-fenichel replied:
It's the padding in the direction perpendicular to the scrollbar, just so that it doesn't bump up against any borders.

picklesrus commented on core/flyout.js:
```diff
@@ -209,46 +250,74 @@ Blockly.Flyout.prototype.getMetrics_ = function() {
+  } else {
+    var viewHeight = this.height_ - 2 * this.SCROLLBAR_PADDING;
+    var viewWidth = this.width_;
+    if (!this.RTL) {
```
Should this be based on the toolbar position enum now instead of rtl?

picklesrus commented on core/flyout.js:
```diff
@@ -209,46 +250,74 @@ Blockly.Flyout.prototype.getMetrics_ = function() {
+    contentTop: optionBox.y,
+    contentLeft: 0,
+    absoluteTop: absoluteTop,
+    absoluteLeft: this.horizontalLayout ? 0 : this.SCROLLBAR_PADDING
```
should this be horizontalLayout ? this.scrollbar_padding : 0 ? or do I just need more coffee?

rachel-fenichel replied:
I think you need more coffee.  The scrollbar is on the bottom in horizontal mode, so we don't need space for it on the left.

picklesrus replied:
Ah, yes. that makes sense.  But then, it used to be 0 in vertical mode so that part is changing. So then the question becomes is this change intentional?

picklesrus commented on core/flyout.js:
```diff
@@ -259,47 +328,147 @@ Blockly.Flyout.prototype.position = function() {
   }
-  var edgeWidth = this.width_ - this.CORNER_RADIUS;
-  if (this.RTL) {
+  var edgeWidth = this.horizontalLayout_ ? metrics.viewWidth : this.width_;
```
is the difference between viewWidth and this.width the scrollbar padding?  And is scrollbar padding applied on the sides of the flyout without the scrollbars?  Sorry for being so confused about the meaning of  SCROLLBAR_PADDING.

picklesrus commented on core/flyout.js:
```diff
@@ -362,18 +535,7 @@ Blockly.Flyout.prototype.hide = function() {
-    goog.dom.removeNode(rect);
-  }
-  this.buttons_.length = 0;
+  this.clearOldBlocks_();
```
This is another thing that could be in a separate pull request since it seems pretty independent. But then, I wonder if it is possible to just call workspace.clear() or workspace.dispose() instead.

rachel-fenichel replied:
Currently moving this to a new pull request with some of the other basic decomposition that I did.  Will update with that number as well.

rachel-fenichel replied:
#342

picklesrus commented on core/flyout.js:
```diff
@@ -260,28 +329,46 @@ Blockly.Flyout.prototype.position = function() {
+  // The blocks need to be visible in order to be laid out and measured
+  // correctly, but we don't want the flyout to show up until it's properly
+  // sized.  Opacity is set to zero in show().
+  this.svgGroup_.style.opacity = 1;
```
Is this left over from a merge?

picklesrus replied:
Ah, I hadn't gotten to show yet.  What happens when you click on something that is transparent? I think want the event to go through it to the thing below, but will it?  My first thought would have been to use visiblity:hidden instead of opacity, but my thoughts on these things come from old internet times so might be outdated.

rachel-fenichel replied:
It shouldn't stay transparent long enough to be clickable.

picklesrus commented on core/flyout.js:
```diff
@@ -427,9 +581,14 @@ Blockly.Flyout.prototype.show = function(xmlList) {
       this, deselectAll));
 
-  this.width_ = 0;
+  if (this.horizontalLayout_) {
```
I realize this just mirrors the width_ = 0 that was already there, but do you know why we set the width to zero when the mouse wheel is scrolling?

rachel-fenichel replied:
This isn't during a scroll.  It's in show().  It gets reset here and then set in reflow().

picklesrus replied:
Argh. I wish this thing never collapsed function definitions.  It would be much easier for me to read if I could always glance up and see what function the diff was in.

picklesrus commented on core/flyout.js:
```diff
@@ -427,9 +581,14 @@ Blockly.Flyout.prototype.show = function(xmlList) {
+  }
   this.reflow();
 
+  this.offsetHorizontalRtlBlocks(this.workspace_.getTopBlocks(false));
```
why do we need this if there is no movement on mouse wheel in the horizontal case?

rachel-fenichel replied:
Also in show(), not onMouseWheel

picklesrus commented on core/flyout.js:
```diff
@@ -446,7 +605,13 @@ Blockly.Flyout.prototype.show = function(xmlList) {
+  // The blocks need to be visible in order to be laid out and measured
+  // correctly, but we don't want the flyout to show up until it's properly
+  // sized.  Opacity is reset at the end of position().
+  this.svgGroup_.style.opacity = 0;
```
How come this wasn't an issue for the vertical case?

rachel-fenichel replied:
In the vertical case we didn't care about how much space the blocks took up until after reflow.  We cared about how much space they took up for setting the background path and also when we measured for scrolling/resizing.  In the horizontal rtl case we need to measure them correctly on the first pass so we can offset them such that the last block is near (0, 0).

rachel-fenichel replied:
This may have been a "fix" for an issue with a different root cause.  I'll take it out and see if anything breaks.

picklesrus commented on core/flyout.js:
```diff
@@ -446,7 +605,13 @@ Blockly.Flyout.prototype.show = function(xmlList) {
+  this.svgGroup_.style.opacity = 0;
+  this.svgGroup_.style.display = 'block';
+
+  var cursorX = margin / this.workspace_.scale + Blockly.BlockSvg.TAB_WIDTH;
```
Can you explain this?  And does it matter if the blocks are RTL or not since the tab flips sides?

rachel-fenichel replied:
Scale should matter for both x and y, and tab width should only be added in the rtl case.

Updated.

picklesrus commented on core/flyout.js:
```diff
@@ -498,6 +670,25 @@ Blockly.Flyout.prototype.clearOldBlocks_ = function() {
+ * Delete blocks and background buttons from a previous showing of the flyout.
+ * @private
+ */
+Blockly.Flyout.prototype.clearOldBlocks_ = function() {
```
Looks like you have 2 of these functions now

rachel-fenichel replied:
Deleted.

picklesrus commented on core/flyout.js:
```diff
@@ -739,36 +889,67 @@ Blockly.Flyout.prototype.createBlockFunc_ = function(originBlock) {
+  // account.
+  if (this.toolboxPosition_ == Blockly.TOOLBOX_AT_RIGHT) {
+    scrollX = workspace.getMetrics().viewWidth - this.width_;
+    scale = workspace.scale;
```
what's the difference between workspace_.scale and workspace.scale?  Practically, they should be the same since zooming changes both?  Just want to make sure the sure the different uses of workspace.scale and this.workspace_.scale are intentional both here and when you're calculating Y.  I don't currently understand why the toolbox on the right and toolbox at bottom would want to use a different scale than left and top.

rachel-fenichel replied:
Good point.  This was originally a static method (effectively) so workspace was passed in rather than accessing it from flyout (also previously passed in).  I can now just access this.workspace_ directly.

rachel-fenichel replied:
I take that back--the workspaces are different.  Workspace is the target workspace (where the block is being placed) and this.workspace_ is the workspace in the flyout.  I can still get rid of workspace as an argument though, and I'm replacing with with a local variable targetWorkspace to make it clear which workspace is which.

With the current state of blockly the two scales should always be the same, but in theory it's possible to do zoom independently in either workspace.

rachel-fenichel replied:
> I don't currently understand why the toolbox on the right and toolbox at bottom would want to use a different scale than left and top.

At left and top, (0, 0) in the flyout matches (0, 0) in the main workspace, so the only transformation we need to apply is if there was a scroll in the flyout.

In the case of right and bottom, the block would be placed relative to the top left of the main workspace (0, 0) instead of under the mouse, so we do an additional transformation by the difference between main workspace (0, 0) and flyout workspace (0, 0).

picklesrus commented on core/flyout.js:
```diff
@@ -799,13 +980,23 @@ Blockly.Flyout.prototype.getClientRect = function() {
+        BIG_NUM + height);
+  } else if (this.toolboxPosition_ == Blockly.TOOLBOX_AT_BOTTOM) {
+    return new goog.math.Rect(-BIG_NUM, y, BIG_NUM * 2,
+        BIG_NUM + height);
```
Just want to double check that this one is right after you fixed the toolbox at bottom overlapping bug.  I think yes since the y coordinate looks right to me, but checking it works in your mulitplayground seems worth doing.

rachel-fenichel replied:
Confirmed that it does.

picklesrus commented on core/flyout.js:
```diff
@@ -833,3 +1024,147 @@ Blockly.Flyout.terminateDrag_ = function() {
+ * Compute width of flyout.  Position button under each block.
+ * For RTL: Lay out the blocks right-aligned.
+ * @param {!Array<!Blockly.Block>} blocks The blocks to reflow.
+ * @param {boolean} opt_skip_resize Possibly skip resizing, since sometimes a
```
I don't see this param below

picklesrus commented on core/flyout.js:
```diff
@@ -833,3 +1024,147 @@ Blockly.Flyout.terminateDrag_ = function() {
+ * @param {boolean} opt_skip_resize Possibly skip resizing, since sometimes a
+ *    resize will be called immediately anyway.
+ */
+Blockly.Flyout.prototype.reflowVertical = function(blocks) {
```
I know this code isn't new, but maybe you can explain it to me in person?  It'll help me review the reflowHorizontal function better. I'd also like to make sure I understand why it is so complicated (and costly) to calculate the size of the flyout.

picklesrus commented on core/flyout.js:
```diff
@@ -833,3 +1024,147 @@ Blockly.Flyout.terminateDrag_ = function() {
+
+/**
+ * Reflow blocks and their buttons.
+ * @param {boolean} opt_skip_resize Possibly skip resizing, since sometimes a
```
does this param belong?

rachel-fenichel replied:
Nope.  Removed, here and above.

picklesrus commented on core/flyout.js:
```diff
@@ -833,3 +1024,147 @@ Blockly.Flyout.terminateDrag_ = function() {
+ * @param {!Array<!Blockly.Block>} blocks The blocks to repositions.
+ */
+Blockly.Flyout.prototype.offsetHorizontalRtlBlocks = function(blocks) {
+  if (this.horizontalLayout_ && this.RTL) {
```
I still don't quite follow this part. I think I understand that you want to lay it out and then shift them all to the other side.  

Side question: Could you avoid the getBBox call if you used the x position of the first block instead?

rachel-fenichel replied:
I want to lay them all out and then shift to the other side, but I don't know how far to shift it over unless I know the view width--it defaults to zero if I don't know it.

I think you mean x position of last block, since it's rtl.  I think that would work.

picklesrus commented on core/options.js:
```diff
@@ -85,7 +105,10 @@ Blockly.Options = function(options) {
   }
 
-  this.RTL = !!options['rtl'];
+  var enableRealtime = !!options['realtime'];
```
why is this realtime stuff here?

rachel-fenichel replied:
Removed.  Maybe a bad merge.

picklesrus commented on core/scrollbar.js:
```diff
@@ -306,9 +306,9 @@ Blockly.Scrollbar.prototype.resizeHorizontal_ = function(hostMetrics) {
   if (this.pair_ && this.workspace_.RTL) {
-    this.xCoordinate += hostMetrics.absoluteLeft +
-        Blockly.Scrollbar.scrollbarThickness;
+    this.xCoordinate += Blockly.Scrollbar.scrollbarThickness;
```
Can you remind me why absoluteLeft is gone? Wasn't this what shifted the position of the scrollbar to the beginning (ie. right) in rtl mode?  Although, now that I look at the existing blockly playground, in rtl mode the scrollbar starts at the right in both RTL and LTR.

rachel-fenichel replied:
I think this was a mistake before, but usually by a small enough number that it didn't matter.

The scrollbar doesn't start at the beginning in RTL, as you noticed.  The point of this offset is to provide the little white square that neither scrollbar will go into, at the corner where they meet.  So it only needs to move over by the thickness of the other scrollbar.  In LTR mode this is accomplished by capping how far it can move over, but here it needs the correct start position to be set.

Previously absoluteLeft has been < 5 so readding it was fine.  No it can be much larger, due to having a vertical toolbox on the left in RTL mode.  Adding it again means that the scrollbar can scroll offscreen to the right but cannot scroll all the way to the left side of the workspace.

picklesrus commented on core/flyout.js:
```diff
@@ -733,42 +851,75 @@ Blockly.Flyout.prototype.createBlockFunc_ = function(originBlock) {
+  // mode.
+  var scrollX = this.workspace_.scrollX;
+  var scale = this.workspace_.scale;
+  xyOld.x += scrollX / scale - scrollX;
```
maybe just put xyOld.x += after the if (toolbox_at_right)?  Looks like the same calculation in both places once you've set scrollX and scale.
Also, why does the right vs left toolbox affect which workspace's scale you use? Shouldn't they both use the same one?

picklesrus replied:
I dunno what I was thinking when I was looking at this code yesterday.  I think I glazed over all the += as =. which had me thoroughly confused. Sorry about that.

picklesrus commented on core/flyout.js:
```diff
@@ -733,42 +851,75 @@ Blockly.Flyout.prototype.createBlockFunc_ = function(originBlock) {
+  // (0, 0) in the main workspace.  Offset to take that into account.
+  if (this.toolboxPosition_ == Blockly.TOOLBOX_AT_BOTTOM) {
+    scrollY = targetWorkspace.getMetrics().viewHeight - this.height_;
+    scale = targetWorkspace.scale;
```
same comment as above

picklesrus commented on core/flyout.js:
```diff
@@ -459,10 +612,17 @@ Blockly.Flyout.prototype.layoutBlocks_ = function(blocks, gaps, margin) {
+    }
+    block.moveBy((this.horizontalLayout_ && this.RTL) ? -cursorX : cursorX,
+        cursorY);
+    if (this.horizontalLayout_) {
```
what does this if block do? looks like cursorX and cursorY aren't used below here? OR I can't find a stupid expandy thing....

picklesrus replied:
Cancel that previous comment.  Git's diff and I are at odds with each other.  Thanks @rachel-fenichel for your patience.

NeilFraser commented on core/flyout.js:
```diff
@@ -193,14 +215,33 @@ Blockly.Flyout.prototype.dispose = function() {
 
 /**
+ * Get the width of the flyout.
+ * @return {number} the width of the flyout.
```
Uppercase 'The'.  And again below.

rachel-fenichel replied:
Done.

NeilFraser commented on core/toolbox.js:
```diff
@@ -50,15 +51,81 @@ Blockly.Toolbox = function(workspace) {
+    this.treeSeparatorConfig_['cssTreeRow'] =
+        'blocklyTreeSeparatorHorizontal' +
+        (workspace.RTL ?
+        ' blocklyHorizontalTreeRtl' : ' blocklyHorizontalTree');
```
If you move the common leading space from both strings here to after  'blocklyTreeSeparatorHorizontal' you can save one byte!

rachel-fenichel replied:
Done.

NeilFraser commented on core/toolbox.js:
```diff
@@ -153,6 +203,22 @@ Blockly.Toolbox.prototype.dispose = function() {
 
 /**
+ * Get the width of the toolbox.
+ * @return {number} the width of the toolbox.
```
Captitalize 'The' here and below.

rachel-fenichel replied:
Done

NeilFraser commented on core/toolbox.js:
```diff
@@ -164,18 +230,31 @@ Blockly.Toolbox.prototype.position = function() {
+    if (this.toolboxPosition == Blockly.TOOLBOX_AT_RIGHT) {  // Right
+      treeDiv.style.left =
+          (svgPosition.x + svgSize.width - treeDiv.offsetWidth) + 'px';
+    } else { // Left
```
Add a second leading space before the comment.  To match the other three comments above.
(This is a style rule for Python at Google, we don't have a rule one way or another for JS.  But the codebase is consistent with a double space before an inline comment.)

rachel-fenichel replied:
Done.

NeilFraser commented on core/toolbox.js:
```diff
@@ -313,16 +394,26 @@ Blockly.Toolbox.prototype.getClientRect = function() {
+  } else if (this.toolboxPosition == Blockly.TOOLBOX_AT_TOP) {
+    return new goog.math.Rect(-BIG_NUM, -BIG_NUM, 2 * BIG_NUM,
+        BIG_NUM + y + height);
+  } else { // Bottom
```
Same here.

rachel-fenichel replied:
Done

NeilFraser commented:
RTL mutators are broken as of this PR.
<img width="230" alt="rtl mutator" src="https://cloud.githubusercontent.com/assets/250480/15234559/c02408b0-1867-11e6-8698-a74e75d23e23.png">

rachel-fenichel commented:
Nooooooooooo

On Thu, May 12, 2016 at 5:34 PM Neil Fraser <notifications@github.com>
wrote:

> RTL mutators are broken as of this PR.
> [image: rtl mutator]
> <https://cloud.githubusercontent.com/assets/250480/15234559/c02408b0-1867-11e6-8698-a74e75d23e23.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/google/blockly/pull/291#issuecomment-218923575>
>

NeilFraser commented:
RTL non-closing horizontal flyouts (no categories) start scrolled all the way to the left.  They should be scrolled to the start which is at the right.  The behaviour with categories is correct.

rachel-fenichel commented:
Is it the flyout that is scrolled or the non flyout workspace inside the
mutator?

Is that in RTL, toolbox at end?

On Thu, May 12, 2016 at 5:55 PM Neil Fraser <notifications@github.com>
wrote:

> RTL non-closing horizontal flyouts (no categories) start scrolled all the
> way to the left. They should be scrolled to the start which is at the
> right. The behaviour with categories is correct.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/google/blockly/pull/291#issuecomment-218926170>
>

NeilFraser commented on core/flyout.js:
```diff
@@ -328,6 +473,10 @@ Blockly.Flyout.prototype.scrollToStart = function() {
  * @private
  */
 Blockly.Flyout.prototype.wheel_ = function(e) {
+  // Don't scroll sideways.
```
Why not?  Many mice have wheels that can scroll in the horizontal plane.  For horizontal scrolling use e.deltaX.

rachel-fenichel replied:
Done.

NeilFraser commented:
RTL flyout in main workspace.
<img width="242" alt="rtl scroll" src="https://cloud.githubusercontent.com/assets/250480/15234985/604a1d26-186c-11e6-8d4e-7b6b76a15341.png">

NeilFraser commented:
Otherwise, this looks really clean.  Both style and functionality.  I'm impressed.

rachel-fenichel commented:
Mutator workspace and RTL horizontal flyout fixed!

rachel-fenichel commented:
And horizontal scrolling done.  PTAL.  So excited for this to land!

NeilFraser commented:
LGTM
